config/src/main/java/com/typesafe/config/impl/ConfigReference.java

```diff
@@ -88,29 +90,40 @@ protected boolean ignoresFallbacks() {
                 v = result.value;
                 newContext = result.context;
             } else {
-                v = null;
+                v = getFallbackSubst(context);
             }
         } catch (NotPossibleToResolve e) {
             if (ConfigImpl.traceSubstitutionsEnabled())
                 ConfigImpl.trace(newContext.depth(),
                         "not possible to resolve " + expr + ", cycle involved: " + e.traceString());
             if (expr.optional())
-                v = null;
+                v = getFallbackSubst(context);
             else
                 throw new ConfigException.UnresolvedSubstitution(origin(), expr
                         + " was part of a cycle of substitutions involving " + e.traceString(), e);
         }
 
         if (v == null && !expr.optional()) {
-            if (newContext.options().getAllowUnresolved())
-                return ResolveResult.make(newContext.removeCycleMarker(this), this);
-            else
-                throw new ConfigException.UnresolvedSubstitution(origin(), expr.toString());
+            throw new ConfigException.UnresolvedSubstitution(origin(), expr.toString());
```

Comment:
wasn't clear to me here why the code for checking `getAllowUnresolved` was removed; I see part of it is now in getFallbackSubst but not the removeCycleMarker